File: Leetcode/MinimumDeletionstoMakeStringBalanced.rs
// https://leetcode.com/problems/minimum-deletions-to-make-string-balanced

struct Solution {}

impl Solution {
    pub fn minimum_deletions(s: String) -> i32 {
        let s = s.into_bytes();

        let mut pref_b = 0;
        let mut suf_a = s.iter().filter(|&x| *x == b'a').count();
        let mut ans = suf_a.min(s.len() - suf_a);

        for c in s.into_iter() {
            if c == b'a' {
                suf_a -= 1;
            } else {
                pref_b += 1;
            }

            ans = ans.min(pref_b + suf_a);
        }

        ans as i32
    }
}

File: Leetcode/MinimumDeletionstoMakeStringBalanced.cpp
// https://leetcode.com/problems/minimum-deletions-to-make-string-balanced

class Solution {
   public:
    int minimumDeletions(string s) {
        const auto an = count(s.cbegin(), s.cend(), 'a');
        const auto bn = count(s.cbegin(), s.cend(), 'b');

        size_t ans = min(an, bn);
        size_t a = 0;
        size_t b = 0;
        for (size_t i = 0; i < s.size(); ++i) {
            a += s[i] == 'a';
            b += s[i] == 'b';

            ans = min(ans, an - a + b);
        }

        return ans;
    }
};